docs/testplan/IPv4-Port-Based-DHCP-Server-test-plan.md

```diff
@@ -0,0 +1,258 @@
+
+## Scope
+
+### Test Scenario
+
+The tests will include:
+
+1. Configuration test
+   1. Add related configuration into CONFIG_DB and then verify configuration and process running status.
+   2. Update related tables in CONFIG_DB to see whether configuration for DHCP Server change too.
+2. Functionality test
+   1. Check whether dhcrelay in dhcp_relay can foward DHCP packets between client and dhcp_server container.
+   2. Check whether dhcp_server container can reply DHCP reply packets as expected.
+   3. Verify in multi-vlan scenario.
+
+### Supported Topology
+
+Base dhcp_server functionality tests (test module [#1](#test-module-1-test_dhcp_serverpy) [#2](#test-module-2-test_dhcp_server_multi_vlanpy) [#3](#test-module-3-test_dhcp_server_stresspy)) are supported on mx topology, smart switch related test (test module [#4](#test-module-4-test_dhcp_server_smart_switchpy)) is supported on t1-smartswitch topology (A new topology on real smart switch testbed).
+
+## Test Case
+
+### Test Module #1 test_dhcp_server.py
+
+#### Port Based Common setup
+
+* Check whether dhcrelay process running as expected (Original dhcp_relay functionality).
+* Enable dhcp_server feature, and then use CLI to add DHCP Server configuration.
+
+#### Port Based Common teardown
+
+* Disable dhcp_server feature, and then check whether dhcrelay process running as expected (Original dhcp_relay functionality).
+* Config reload, remove dhcp_server container.
+
+#### test_dhcp_server_port_based_assignment_single_ip
+
+* **Test objective**
+
+   To test port based single ip assign.
+
+* **Test detail**
+
+  * Send DHCP discover packets from ptf, check whether configured port receive DHCP offer packet and no-configured ports don't receive. Need to check netmask / gateway / lease_time / yiaddr.
+  * Send DHCP request packets from ptf, check whether configured port receive DHCP ack packet and no-configured ports don't receive. Need to check netmask / gateway / lease_time / yiaddr. Besides, check lease via show CLI to make sure lease is correct.
+  * For renew scenario, send DHCP request packets from ptf, check whether configured port receive DHCP ack packet and no-configured ports don't receive. Need to check netmask / gateway / lease_time / yiaddr. Besides, check lease via show CLI to make sure lease is correct.
+  * Send DHCP release packets from ptf, check whether lease release via lease file inside dhcp_server container.
+
+#### test_dhcp_server_port_based_assignment_range
+
+* **Test objective**
+
+   To test port based range ip assign.
+
+* **Setup**
+
+  Add range and bind range via CLI.
+
+* **Teardown**
+
+  Unbind range and del range via CLI.
+
+* **Test detail**
+
+  * Always send packets from 1 ptf port with different client mac, process of sending and verifying can refer to `test_dhcp_server_port_based_assignment_single_ip`.
```

Comment:
Can we verify below case as well:
dhclient numbers larger than assignable IP addresses

Reply:
Already included in next line `When IPs in range are all used, new client cannot get IP.`